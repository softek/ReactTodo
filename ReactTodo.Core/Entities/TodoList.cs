using Ardalis.GuardClauses;
using ReactTodo.SharedKernel;

namespace ReactTodo.Core.Entities;

public class TodoList : BaseEntity, IAggregateRoot
{
    private List<TodoTask> _items = [];

#pragma warning disable CS8618
    protected TodoList() {/* Required for EF */}
#pragma warning restore CS8618

    public TodoList(string name) =>
        Name = Guard.Against.NullOrEmpty(name);

    public string Name { get; private set; }

    public IEnumerable<TodoTask> Tasks => _items.AsReadOnly();

    public void AddItem(TodoTask task) =>
        _items.Add(Guard.Against.Null(task));

    public void RemoveItem(TodoTask task) =>
        _items.Remove(Guard.Against.Null(task));
}
