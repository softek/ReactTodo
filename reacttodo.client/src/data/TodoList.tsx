export type TodoList = {
  name: string;
  id: string;
};
