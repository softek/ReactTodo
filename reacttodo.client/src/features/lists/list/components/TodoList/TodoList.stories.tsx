import type { Meta, StoryObj } from '@storybook/react';
import {
  expect, fn, userEvent, waitFor, within,
} from '@storybook/test';
import { TodoList } from './TodoList';

const meta = {
  title: 'Features/TodoList',
  component: TodoList,
  parameters: {
    layout: 'centered',
  },
  argTypes: {
  },
  args: {
    onDeleteTask: fn(),
    onUpdateTask: fn(),
    onCreateTask: fn(),
  },
} satisfies Meta<typeof TodoList>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  args: {
    name: 'My List',
    tasks: [
      { id: '1', name: 'Task 1', isComplete: false },
      { id: '2', name: 'Task 2', isComplete: true },
    ],
  },
  play: async ({ args, canvasElement }) => {
    const canvas = within(canvasElement);

    const box = canvas.getByLabelText('Task 1');
    await expect(box).toBeInTheDocument();
    userEvent.click(box);
    await waitFor(() => expect(args.onUpdateTask).toHaveBeenCalledWith({ id: '1', name: 'Task 1', isComplete: true }));

    const deleteButton = canvas.getByRole('button', { name: /Delete/i });
    await expect(deleteButton).toBeInTheDocument();
    await userEvent.click(deleteButton);
    await waitFor(() => expect(args.onDeleteTask).toHaveBeenCalledWith('2'));

    const editButton = canvas.getByRole('button', { name: /Edit/i });
    await expect(editButton).toBeInTheDocument();
    await userEvent.click(editButton);
    const editBox = canvas.getByDisplayValue('Task 1');
    await expect(editBox).toBeInTheDocument();
    await userEvent.clear(editBox);
    await userEvent.type(editBox, 'Renamed Task 1');
    const saveButton = canvas.getByRole('button', { name: /Save/i });
    await userEvent.click(saveButton);
    await waitFor(() => expect(args.onUpdateTask).toHaveBeenCalledWith({ id: '1', name: 'Renamed Task 1', isComplete: false }));

    const createButton = canvas.getByRole('button', { name: /Create Task/i });
    const newTaskBox = canvas.getByPlaceholderText('New task');
    await expect(createButton).toBeInTheDocument();
    await userEvent.type(newTaskBox, 'New Task Name');
    await userEvent.click(createButton);
    await waitFor(() => expect(args.onCreateTask).toHaveBeenCalledWith('New Task Name'));
    await expect(newTaskBox).toHaveValue('');
  },
};
