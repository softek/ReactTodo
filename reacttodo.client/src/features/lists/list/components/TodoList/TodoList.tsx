import React from 'react';
import { ListGroup } from 'reactstrap';
import { TodoTask } from '../../../../../data';
import { TodoListItem } from '../TodoListItem';
import { NewItem } from '../NewItem';

export type TodoLisProps = {
  name: string;
  tasks: TodoTask[];
  onDeleteTask: (id: string) => void;
  onUpdateTask: (task: TodoTask) => void;
  onCreateTask: (name: string) => void;
}

export function TodoList({
  name, tasks, onDeleteTask, onUpdateTask, onCreateTask,
}: TodoLisProps) {
  return (
    <div>
      <h1>{name}</h1>
      <NewItem onCreateTask={(task) => onCreateTask(task)} />
      <ListGroup>
        {tasks.map((task) => (
          <TodoListItem
            key={task.id}
            task={task}
            onDelete={() => onDeleteTask(task.id)}
            onToggle={() => onUpdateTask({ ...task, isCompleted: !task.isCompleted })}
            onRename={(newName) => onUpdateTask({ ...task, name: newName })}
          />
        ))}
      </ListGroup>
    </div>
  );
}
