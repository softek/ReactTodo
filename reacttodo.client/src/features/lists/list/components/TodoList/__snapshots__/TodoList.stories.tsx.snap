// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Features/TodoList Default play-test 1`] = `
<div class="container-fluid">
  <h1>
    My List
  </h1>
  <form>
    <div class="p-0 d-flex mb-4">
      <input placeholder="New task"
             name="name"
             type="text"
             class="form-control"
             aria-invalid="false"
             value
      >
      <button type="button"
              class="btn btn-success"
              aria-label="Create Task"
      >
        Create Task
      </button>
    </div>
  </form>
  <ul class="list-group">
    <li class="list-group-item">
      <div class="form-check form-check-inline">
        <input id="task-completion-1"
               type="checkbox"
               class="form-check-input"
        >
        <label for="task-completion-1"
               class="form-check-label"
        >
          Task 1
        </label>
      </div>
      <button type="button"
              class="btn btn-secondary"
              aria-label="Edit"
      >
        Edit
      </button>
    </li>
    <li class="list-group-item">
      <div class="form-check form-check-inline">
        <input id="task-completion-2"
               type="checkbox"
               class="form-check-input"
               checked
        >
        <label for="task-completion-2"
               class="text-decoration-line-through form-check-label"
        >
          Task 2
        </label>
      </div>
      <button type="button"
              class="btn btn-danger"
              aria-label="Delete"
      >
        Delete
      </button>
    </li>
  </ul>
</div>
`;
